volumes: make Terraform import work

okrause opened the pull request:
This PR is a possible solution for volume import (see issue #33 ).

It implements approach number 1. It is the most simply solution and should work well, but will fail for the very unlikely corner case that more than one volume hold a volume with the requested volume UUID.

A proper importer which can also handle cases where multiple volumes with the same ID are found would be the cleaner solution.

This is how to use it:

1. Create an empty volume resource (volume.tf), e.g.:
```tf
resource "netapp-gcp_volume" "myvolume" {
}

```
2. Lookup the volumeID of the volume you want to import in the UI, e.g 6462b343-262c-e785-dc6c-ca5fcfc7e185.
3. Run the import
`terraform import netapp-gcp_volume.myvolume 6462b343-262c-e785-dc6c-ca5fcfc7e185
`
4. If it worked fine, TF will have imported the state of the volume into the terraform.tfstate file. Use 
`terraform show -no-color > volume.tf` to create the resource definition (overwriting the prior empty definition).
5. Delete the `id` line in volume.tf
6. Run `terraform plan`. If everything went fine, it should run without proposing changes.
7. Voila, you now have the definition in volume.tf and can start managing your volume with TF.

Happy to discuss.

okrause committed:
volumes: make Terraform import work

okrause committed:
Merge branch 'master' into import

okrause commented:
Observation for shared-vpcs:
Importing and subsequent management of volumes with above method works fine.
Recreating the volume fails, since the TF files (generated via terraform show) are missing the shared_vpc_project_number attribute. It needs to be added manually to the TF file.

lonico commented on gcp/volume.go:
```diff
@@ -184,6 +195,45 @@ func (c *Client) getVolumeByRegion(region string) ([]volumeResult, error) {
 	return volumes, nil
 }
 
+func (c *Client) findRegionbyID(VolumeID string) (volumeResult, error) {
```
should be  findVolumeByIDAcrossRegions  
how expensive an operation is it if we don't give a region?

okrause replied:
To stay API compatible, this extends the current getVolumeByID. In getVolumeByID it calls findRegionByID, if region is not set (which is/was a requirement for getVolumeByID to work). findRegionByID does one additional API call (enumerate all volumes) and filters the list by ID to extract the region. This region information is then used by getVolumeByID to do its job. tl;dr: If no region is provided, we find the missing region by filtering all volumes for given UUID.
Since this extends getVolumeByID and is not substituting it, I don't think findVolumeByIDAcrossRegions would be an appropriate name. Open to discuss.

lonico replied:
this function returns a volume (return resultVolume, nil) not a region.

okrause replied:
Yes, you are correct. I will change the function name.

lonico commented on gcp/volume.go:
```diff
@@ -184,6 +195,45 @@ func (c *Client) getVolumeByRegion(region string) ([]volumeResult, error) {
+	for _, eachVolume := range result {
+		if eachVolume.VolumeID == VolumeID {
+			count = count + 1
+			resultVolume = eachVolume
```
I think it would be better to build a list of volumes.   So if we find more than one, we can give more information of why the import failed.  The user could delete older volumes in other regions for instance.

okrause replied:
Yes, we can extend it to show the conflicting volumes. But, as you said, deleting would be the only way for resolution for the user, which is an undesirable solution.

lonico commented on gcp/volume.go:
```diff
@@ -184,6 +195,45 @@ func (c *Client) getVolumeByRegion(region string) ([]volumeResult, error) {
+	}
+
+	if count > 1 {
+		return volumeResult{}, fmt.Errorf("Found more than one volume for VolumeID: %v", VolumeID)
```
here, I'd like to show more info about all the volumes with the same ID.

okrause commented:
@lonico thank you for the input above.
Thinking about the ways to resolve multiple volumes with the same ID made me think about the actual problem again.

Our actual problem it, that ID is not guaranteed to be unique over all regions. The import command required a unique ID. So how would a unique ID look like? It is actually an (ID, region) tuple. Maybe we should simply require the user to specify both, e.g by requiring an ID in "CVSID:region_name" format. We could parse the components and don't need to guess the users intention.
Drawback is, that we cannot simply use the ID provided by the UI, might break user expectations and therefore need to document it well.

lonico commented:
@okrause I like your idea about "CVSID:region_name".   And we could support both format, if we assume that ":" cannot show in CSVID or region_name.
If we receive "CVSID", we try to find the region, but error out if multiple IDs are found.  In the error message, we can ask for "CVSID:region_name" to resolve the ambiguity.
If we receive "CVSID:region_name" then we know where to look for.

okrause commented:
@lonico Great proposal, like it a lot. It maintains user convenience and terraform import usage consistency, while proving a solution for this super rare corner case. Will update the PR over x-mas. Have a wonderful holiday.

okrause committed:
Implement generic volume filtering

- getAllVolumes is a helper function which returns all volumes in the
  project as a slice
- filterAllVolumes will apply a filter function to the list of all
  volumes and returns a slice of matching volumes
- getVolumeByID now uses filterAllVolumes in case region is not set
  (will happen for terraform import)

okrause committed:
More refactoring in getVolume* functions

- getAllVolumes now takes region argument and is renamed to getVolumes
- refactored getVolumeByRegion and commented it out, since it is not
  used

okrause committed:
getVolumeByID: terraform import improvements

support ID = <volumeID> and ID = <volumeID>:<region> for terraform
import

okrause commented:
Hohohoho, here are coming the x-mas eve changes to terraform import.

A user can continue to import using the standard way (`ID = <volumeID>`), e.g.
`terraform import netapp-gcp_volume.myvolume 1bc88bc6-cc7d-5fe3-3737-8e635fe2f996`.
If only one volume with this ID exists, the import will succeed. If multiple volumes with the ID exist, an error message will instruct the user to use the alternative format `ID = <volumeID>:<region>`, e.g.
`terraform import netapp-gcp_volume.myvolume 1bc88bc6-cc7d-5fe3-3737-8e635fe2f996:us-central1`

I also did some refactoring to reduce redundant code.

Feedback welcome

okrause committed:
terraform import: read back shared_vpc_project_number

When importing volumes on a shared VPC, we have to repopulate
attribute shared_vpc_project_number

okrause committed:
terraform import: make shared_vpc_project_number more robust

Turns out API ((sometimes?)also returns long network path
without shared VPC. Make projectID detection more stable.

wenjun666 commented on gcp/volume.go:
```diff
@@ -158,30 +199,59 @@ func (c *Client) getVolumeByID(volume volumeRequest) (volumeResult, error) {
 	}
+
+	// volumeID is verified by Terraform. If we use ID = <volumeID>:<region>, we need to revert our ID changes
+	if originalID != "" {
+		result.VolumeID = originalID
+	}
```
result.VolumeID is returned by the API, which has the <volumeID> format, is it necessary to set it again here?

okrause replied:
It is necessary for the `ID = <volumeID>:<region>` case. Why?

The user submits an import like 
`terraform import netapp-gcp_volume.myvolume 1bc88bc6-cc7d-5fe3-3737-8e635fe2f996:us-central1`
This means that volumeID comes in as `1bc88bc6-cc7d-5fe3-3737-8e635fe2f996:us-central1`, which is an invalid volumeID from API perspective. In line 153-156 I extract real volumeID and region from it and use it to do the API call.

So far so good. volumeID now contains `1bc88bc6-cc7d-5fe3-3737-8e635fe2f996` and this get returned to Terraform. But now Terraform complains that the ID doesn't match any more, because in its books this import got ID `1bc88bc6-cc7d-5fe3-3737-8e635fe2f996:us-central1`. To make TF happy, I set volumeID back to `1bc88bc6-cc7d-5fe3-3737-8e635fe2f996:us-central1`.

lonico commented on gcp/resource_netapp_gcp_volume.go:
```diff
@@ -469,8 +469,8 @@ func resourceGCPVolumeCreate(d *schema.ResourceData, meta interface{}) error {
 func waitForVolumeCreationComplete(client *Client, volumeRes volumeResult) (volumeResult, error) {
-	waitSeconds := 900	// first volume creation can take 11 minutes
-	threshold := 900 -60	// when to warn
+	waitSeconds := 900    // first volume creation can take 11 minutes
```
It would be nice to avoid space only changes.

okrause replied:
I agree, but issue here is unclean formatting of the existing code.
It was:
```go
	waitSeconds := 900   // first volume creation can take 11 minutes
	threshold := 900 -60 // when to warn
```
My IDE (vscode) runs `go fmt` when saving, which fixed the missing space in "900 -60", which moved the comment one char to the right. Since "go fmt" aligns the comments on subsequent lines, the comment of the previous line was also moved one to the right.

Here my "test case":
```bash
cat main.go 
package main

func main() {
   // 012345678901234567890
   waitSeconds := 900   // first volume creation can take 11 minutes
   threshold := 900 -60 // when to warn
}
➜  ~/tmp/go-test go fmt
main.go
➜  ~/tmp/go-test cat main.go
package main

func main() {
	// 012345678901234567890
	waitSeconds := 900    // first volume creation can take 11 minutes
	threshold := 900 - 60 // when to warn
}
```

Please note, that the comment in the waitSeconds line also moved to align with the threshold line.

I propose we run "go fmt" before every commit. This will avoid subsequent "empty" format changes.

lonico replied:
Agreed.  And we could run go fmt in our pipeline and error out if there is any change.

lonico commented on gcp/volume.go:
```diff
@@ -141,8 +142,48 @@ type mountPoints struct {
+	// terraform import will specify volumeID.
+	// Issue is, that volumeID is unqiue per region, but might exist in different regions in same project.
+	// For that case, we will support a special ID format for terraform import ADDR ID.
+	// ID = <volumeID>:<region>
```
It's a nitpick
Would   <region>:<volumeID>   be more natural?

I don't really have a preference.

okrause replied:
Sorry, I don't understand. Do you mean replace "will" by "would"? Please clarify.

lonico replied:
My comment was lost in translation :)   The `"<xxx>"` are transformed into "<xxx>"

It's a nitpick
Would   `<region>:<volumeID>`   be more natural?

I don't really have a preference.

okrause replied:
Ahh, I understand. You are right, it would be more natural. But I still like to stick with the current way, because it better reflects the workflow. As a suer I would start out by specifying the volumeID only. This will work in nearly all cases. Only for the unlikely case that we really have a volumeID collision, one would add the region. A user would do that by simply get the existing command from command history. The cursor would be already at the end of the line, so adding the region at the end is more convenient.

lonico replied:
good for me.  Thanks!

lonico commented on gcp/volume.go:
```diff
@@ -141,8 +142,48 @@ type mountPoints struct {
+				If this happend while running terraform import, please use \n
+				terraform import ADDR ID, with ID using <volumeID>:<region_name> format`, volume.VolumeID)
+			}
+			return volumeResult{}, fmt.Errorf("getVolumeByID: WTF error with volume ID %s", volume.VolumeID)
```
WTF?

WTF -> Unexpected 


Though I'm not sure it makes sense anyway.   len is  >= 0,  so we exhausted all cases here.

okrause replied:
You are absolutely right. I was simply overly careful and did hope to sneak in an easter egg. Damn, you found it 😄 . I have a change which removes it and also un-mangles the nested conditions.

lonico commented on gcp/volume.go:
```diff
@@ -158,30 +199,59 @@ func (c *Client) getVolumeByID(volume volumeRequest) (volumeResult, error) {
+func (c *Client) filterAllVolumes(f func(volumeResult) bool) ([]volumeResult, error) {
+	filteredVolumes := make([]volumeResult, 0)
+
+	vols, err := c.getVolumes("-")
```
Are we getting all volumes in all regions?    This sounds like an expensive operation.

okrause replied:
This is a helper function for the use cases where we need to get a list or sublist of all volumes. It is used only in the specific cases where this is necessary.
In our code, we have two places where this is needed:
1) getVolumeByRegion, which isn't used anywhere in the code so I commented it out
2) getVolumeByID. Here it is only required in the import case, where the user specified a volume UUID, but no region. We use it to lookup the specified UUID in all regions.

okrause committed:
TF import: state check cleanups
